File: runtime/cpp/titan-ast-runtime-lib/DfaTokenAutomata.cpp
//
// Created by tian wei jun on 2022/11/25 0025.
//

#include "DfaTokenAutomata.h"
#include <sstream>

const int DfaTokenAutomata::eof = -1;

DfaTokenAutomata::DfaTokenAutomata(const TokenDfa *tokenDfa) : dfa(tokenDfa),oneTokenStringBuilder(256) {}

// dfa delete by PersistentObject.tokenDfa
DfaTokenAutomata::~DfaTokenAutomata() = default;

TokensResult *DfaTokenAutomata::buildToken(const std::string *sourceFilePath) {
  ByteBufferedInputStream byteBufferedInputStream;
  auto initResult = byteBufferedInputStream.init(sourceFilePath);
  if (!initResult.isOk) {
    // byteBufferedInputStream初始化错误（可能原因：源文件不存在）
    return TokensResult::generateSourceIoErrorResult(
        new std::string(initResult.msg));
  }
  // tokens delete by caller
  auto *tokens = new std::list<Token *>();
  TokensResult *tokensResult = nullptr;
  while (true) {
    BuildOneTokenMethodResult *buildOneTokenMethodResult =
        buildOneToken(byteBufferedInputStream);
    if (buildOneTokenMethodResult->type == BuildOneTokenMethodResultType::TOKEN) {
      tokens->push_back(buildOneTokenMethodResult->getTokenData());
      delete buildOneTokenMethodResult;
      continue;
    }
    if (buildOneTokenMethodResult->type == BuildOneTokenMethodResultType::ALL_TEXT_HAS_BEEN_BUILT) {
      auto arrayTokens = new std::vector<Token *> (tokens->size(), nullptr);
      int indexOfToken = 0;
      for(auto token : *tokens){
        (*arrayTokens)[indexOfToken++] = token;
      }
      tokensResult = TokensResult::generateOkResult(arrayTokens);
      delete buildOneTokenMethodResult;
      break;
    }
    if (buildOneTokenMethodResult->type == BuildOneTokenMethodResultType::TOKEN_PARSE_ERROR) {
      BuildOneTokenMethodTokenGeneratorErrorData *tokenGeneratorErrorData =
          buildOneTokenMethodResult
              ->getBuildOneTokenMethodTokenGeneratorErrorData();
          auto arrayTokens = new std::vector<Token *> (tokens->size(), nullptr);
          int indexOfToken = 0;
          for(auto token : *tokens){
            (*arrayTokens)[indexOfToken++] = token;
          }
          tokensResult = TokensResult::generateTokenParseErrorResult(
          new TokenParseErrorData(arrayTokens, tokenGeneratorErrorData->start,
                                  tokenGeneratorErrorData->end,
                                  tokenGeneratorErrorData->errorText));
      delete buildOneTokenMethodResult;
      break;
    }
    if (buildOneTokenMethodResult->type == BuildOneTokenMethodResultType::IO_ERROR) {
      tokensResult = TokensResult::generateSourceIoErrorResult(new std::string(
          "read data from file '" + *sourceFilePath + "' error"));
      for (Token *token : *tokens) {
        delete token;
      }
      delete tokens;
      delete buildOneTokenMethodResult;
      break;
    }
    for (Token *token : *tokens) {
      delete token;
    }
    delete tokens;
    delete buildOneTokenMethodResult;
    break;
  };

  return tokensResult;
}

BuildOneTokenMethodResult *DfaTokenAutomata::buildOneToken(
    ByteBufferedInputStream &byteBufferedInputStream) {
  oneTokenStringBuilder.clear();
  int startIndexOfToken = byteBufferedInputStream.nextReadIndex;
  auto readResult = byteBufferedInputStream.read();
  if (!readResult.isOk) {
    return BuildOneTokenMethodResult::generateIoErrorResult();
  }
  int ch = readResult.data;
  if (ch == eof) {// 输入流读取完了
    return BuildOneTokenMethodResult::generateAllTextHasBeenBuiltResult();
  }
  // first terminal state
  const TokenDfaState *firstTerminalState = nullptr;
  const TokenDfaState *currentState = dfa->start;
  while (ch != eof) {
    TokenDfaState *nextState = nullptr;
    auto nextStateIt = currentState->edges.find(ch);
    if (nextStateIt != currentState->edges.end()) {
      nextState = nextStateIt->second;
    }
    oneTokenStringBuilder.append((byte) ch);
    currentState = nextState;
    if (!nextState) {// 不通
      break;
    }
    if (FaStateType::isClosingTag(currentState->type)) {// 找到终态
      firstTerminalState = currentState;
      byteBufferedInputStream.mark();
      break;
    }
    readResult = byteBufferedInputStream.read();
    if (!readResult.isOk) {
      return BuildOneTokenMethodResult::generateIoErrorResult();
    }
    ch = readResult.data;
  }
  if (!firstTerminalState) {
    std::string tokenStr((char *) (oneTokenStringBuilder.buffer),
                         oneTokenStringBuilder.position);
    return BuildOneTokenMethodResult::generateTokenParseErrorResult(
        new BuildOneTokenMethodTokenGeneratorErrorData{
            startIndexOfToken,
            startIndexOfToken + oneTokenStringBuilder.length(), tokenStr});
  }
  // 重复嗅探更高优先级或贪婪
  int lengthOfToken = oneTokenStringBuilder.length();
  // heaviest terminal state
  const TokenDfaState *heaviestTerminalState = firstTerminalState;
  readResult = byteBufferedInputStream.read();
  if (!readResult.isOk) {
    return BuildOneTokenMethodResult::generateIoErrorResult();
  }
  ch = readResult.data;
  // 如果没有文本嗅探了直接跳出循环
  // 如果当前接受状态是acceptWhenFirstArriveAtTerminalState，就直接接受,跳出嗅探循环
  while (ch != eof) {
    TokenDfaState *nextState = nullptr;
    auto nextStateIt = currentState->edges.find(ch);
    if (nextStateIt != currentState->edges.end()) {
      nextState = nextStateIt->second;
    }
    oneTokenStringBuilder.append((byte) ch);
    currentState = nextState;
    if (!nextState) {// 不通
      break;
    }
    if (FaStateType::isClosingTag(
            currentState->type)) {// 找到终态
                                  // 新状态具有更高优先级的，接受终态转移
      bool isHigherPriority =
          currentState->weight > heaviestTerminalState->weight;
      // 相同优先级说明状态是同一个token的终态
      // 如果是贪婪的，则增加识别的字符，接受终态转移
      // 不是贪婪的，则不接受终态转移
      bool isSameAndGreediness =
          heaviestTerminalState->terminal == currentState->terminal && ((TerminalGrammar *) heaviestTerminalState->terminal)->lookaheadMatchingMode == LookaheadMatchingMode::GREEDINESS;
      // 新状态具有更高优先级的，接受状态转移
      if (isHigherPriority || isSameAndGreediness) {
        heaviestTerminalState = currentState;
        lengthOfToken = oneTokenStringBuilder.length();
        byteBufferedInputStream.mark();
      }
      // 新token优先级更低直接被覆盖，不接受替换旧终态
    }
    readResult = byteBufferedInputStream.read();
    if (!readResult.isOk) {
      return BuildOneTokenMethodResult::generateIoErrorResult();
    }
    ch = readResult.data;
  }
  byteBufferedInputStream.reset();
  oneTokenStringBuilder.setPosition(lengthOfToken);
  // build token result
  std::string text((char *) oneTokenStringBuilder.buffer,
                   oneTokenStringBuilder.length());
  auto type = TokenTypeNamespace::getByGrammarAction(
      heaviestTerminalState->terminal->action);
  auto *token = new Token(*heaviestTerminalState->terminal, startIndexOfToken,
                          text, type);
  return BuildOneTokenMethodResult::generateTokenResult(token);
}

BuildOneTokenMethodResult::BuildOneTokenMethodResult(
    BuildOneTokenMethodResultType type, void *data)
    : type(type), data(data) {}
BuildOneTokenMethodResult::~BuildOneTokenMethodResult() {
  switch (type) {
    case BuildOneTokenMethodResultType::
        TOKEN:                                                  // delete data by TokensResult,DfaTokenAutomata::buildToken
    case BuildOneTokenMethodResultType::ALL_TEXT_HAS_BEEN_BUILT:// delete nothing
    case BuildOneTokenMethodResultType::IO_ERROR:               // delete nothing
      // delete nothing
      break;
    case BuildOneTokenMethodResultType::TOKEN_PARSE_ERROR: {
      auto tokenGeneratorErrorData =
          getBuildOneTokenMethodTokenGeneratorErrorData();
      delete tokenGeneratorErrorData;
      break;
    }
  }
}

BuildOneTokenMethodResult *
BuildOneTokenMethodResult::generateTokenResult(Token *data) {
  return new BuildOneTokenMethodResult(BuildOneTokenMethodResultType::TOKEN,
                                       data);
}
Token *BuildOneTokenMethodResult::getTokenData() const {
  return static_cast<Token *>(data);
}

BuildOneTokenMethodResult *
BuildOneTokenMethodResult::generateAllTextHasBeenBuiltResult() {
  return new BuildOneTokenMethodResult(
      BuildOneTokenMethodResultType::ALL_TEXT_HAS_BEEN_BUILT, nullptr);
}

BuildOneTokenMethodResult *
BuildOneTokenMethodResult::generateTokenParseErrorResult(
    BuildOneTokenMethodTokenGeneratorErrorData *data) {
  return new BuildOneTokenMethodResult(
      BuildOneTokenMethodResultType::TOKEN_PARSE_ERROR, data);
}
BuildOneTokenMethodTokenGeneratorErrorData *
BuildOneTokenMethodResult::getBuildOneTokenMethodTokenGeneratorErrorData()
    const {
  return static_cast<BuildOneTokenMethodTokenGeneratorErrorData *>(data);
}

BuildOneTokenMethodResult *BuildOneTokenMethodResult::generateIoErrorResult() {
  return new BuildOneTokenMethodResult(BuildOneTokenMethodResultType::IO_ERROR,
                                       nullptr);
}

File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/DfaTokenAutomata.java
package titan.ast.runtime;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import titan.ast.runtime.AstGeneratorResult.TokenParseErrorData;
import titan.ast.runtime.AstGeneratorResult.TokensResult;

/**
 * 识别token的dfa.
 *
 * @author tian wei jun
 */
public class DfaTokenAutomata implements TokenAutomata {

  final TokenDfa dfa;
  static final int EOF = -1;
  ByteBuffer tokenStringBuilder = new ByteBuffer(256);

  /**
   * 初始化oneTokenStringBuilder、eof、dfa.
   *
   * @param dfa 识别token的 确定有限状态自动机
   */
  public DfaTokenAutomata(TokenDfa dfa) {
    this.dfa = dfa;
  }

  @Override
  public TokensResult buildToken(String sourceFilePath) {
    TokensResult tokensResult = null;
    try (FileInputStream fileInputStream = new FileInputStream(sourceFilePath)) {
      tokensResult = buildToken(fileInputStream);
    } catch (IOException e) {
      tokensResult = TokensResult.generateSourceIoErrorResult(e.getMessage());
    }
    return tokensResult;
  }

  /**
   * 根据输入流构造tokens.
   *
   * @param byteInputStream 将要识别文本的输入流.
   * @return
   */
  @Override
  public TokensResult buildToken(InputStream byteInputStream) {
    // init
    ByteBufferedInputStream byteBufferedInputStream = new ByteBufferedInputStream(byteInputStream);
    TokensResult tokensResult = null;
    List<Token> tokens = new LinkedList<>();
    while (true) { // build one token
      BuildOneTokenMethodResult buildOneTokenMethodResult = null;
      try {
        buildOneTokenMethodResult = buildOneToken(byteBufferedInputStream);
      } catch (IOException e) {
        tokensResult = TokensResult.generateSourceIoErrorResult(e.getMessage());
        break;
      }
      BuildOneTokenMethodResultType buildOneTokenMethodResultType =
          buildOneTokenMethodResult.getType();
      if (buildOneTokenMethodResultType
          == BuildOneTokenMethodResultType.TOKEN) { // has built one token,add it
        tokens.add(buildOneTokenMethodResult.getTokenData());
        continue;
      }

      if (buildOneTokenMethodResultType == BuildOneTokenMethodResultType.ALL_TEXT_HAS_BEEN_BUILT) {
        ArrayList<Token> tokensArrayList = new ArrayList<Token>(tokens.size());
        tokensArrayList.addAll(tokens);
        tokensResult = TokensResult.generateOkResult(tokensArrayList);
        break;
      }
      if (buildOneTokenMethodResultType == BuildOneTokenMethodResultType.TOKEN_PARSE_ERROR) {
        BuildOneTokenMethodTokenGeneratorErrorData tokenGeneratorErrorData =
            buildOneTokenMethodResult.getBuildOneTokenMethodTokenGeneratorErrorData();
        ArrayList<Token> tokensArrayList = new ArrayList<Token>(tokens.size());
        tokensArrayList.addAll(tokens);
        tokensResult =
            TokensResult.generateTokenParseErrorResult(
                new TokenParseErrorData(
                    tokensArrayList,
                    tokenGeneratorErrorData.start,
                    tokenGeneratorErrorData.end,
                    tokenGeneratorErrorData.errorText));
        break;
      }
      break;
    }
    return tokensResult;
  }

  /**
   * .
   *
   * @return 返回null，表示输入流读取完了
   */
  private BuildOneTokenMethodResult buildOneToken(ByteBufferedInputStream byteBufferedInputStream)
      throws IOException {
    int startIndexOfToken = byteBufferedInputStream.nextReadIndex;
    int ch = byteBufferedInputStream.read();
    if (ch == EOF) { // 输入流读取完了
      return BuildOneTokenMethodResult.generateAllTextHasBeenBuiltResult();
    }
    tokenStringBuilder.clear();
    // first terminal state
    TokenDfaState firstTerminalState = null;
    TokenDfaState currentState = dfa.start;
    while (ch != EOF) {
      TokenDfaState nextState = currentState.edges.get(ch);
      tokenStringBuilder.append(ch);
      currentState = nextState;
      if (null == nextState) { // 不通
        break;
      }
      if (FaStateType.isClosingTag(currentState.type)) { // 找到终态
        firstTerminalState = currentState;
        byteBufferedInputStream.mark();
        break;
      }
      ch = byteBufferedInputStream.read();
    }
    if (null == firstTerminalState) { // does not match any token
      return BuildOneTokenMethodResult.generateTokenParseErrorResult(
          new BuildOneTokenMethodTokenGeneratorErrorData(
              startIndexOfToken,
              startIndexOfToken + tokenStringBuilder.length(),
              tokenStringBuilder.toString()));
    }
    // 重复嗅探更高优先级或贪婪
    int lengthOfToken = tokenStringBuilder.length();
    // heaviest terminal state
    TokenDfaState heaviestTerminalState = firstTerminalState;

    // 如果没有文本嗅探了直接跳出循环
    // 贪婪或者嗅探高优先级
    ch = byteBufferedInputStream.read();
    while (ch != EOF) {
      TokenDfaState nextState = currentState.edges.get(ch);
      tokenStringBuilder.append(ch);
      currentState = nextState;
      if (null == nextState) { // 不通
        break;
      }
      if (FaStateType.isClosingTag(currentState.type)) { // 找到终态
        // 新状态具有更高优先级的，接受终态转移
        boolean isHigherPriority = currentState.weight > heaviestTerminalState.weight;
        // 相同优先级说明状态是同一个token的终态
        // 如果是贪婪的，则增加识别的字符，接受终态转移
        // 不是贪婪的，则不接受终态转移
        boolean isSameAndGreediness =
            heaviestTerminalState.terminal == currentState.terminal
                && ((TerminalGrammar) heaviestTerminalState.terminal).lookaheadMatchingMode
                    == LookaheadMatchingMode.GREEDINESS;
        if (isHigherPriority || isSameAndGreediness) {
          heaviestTerminalState = currentState;
          lengthOfToken = tokenStringBuilder.length();
          byteBufferedInputStream.mark();
        }
        // 新token优先级更低直接被覆盖，不接受替换旧终态
      }
      ch = byteBufferedInputStream.read();
    }
    byteBufferedInputStream.reset();
    tokenStringBuilder.setPosition(lengthOfToken);

    // build token result
    Token token = new Token(startIndexOfToken);
    token.text = tokenStringBuilder.toString();
    token.terminal = heaviestTerminalState.terminal;
    token.type = TokenType.getByGrammarAction(heaviestTerminalState.terminal.action);
    return BuildOneTokenMethodResult.generateTokenResult(token);
  }

  private enum BuildOneTokenMethodResultType {
    TOKEN,
    ALL_TEXT_HAS_BEEN_BUILT,
    TOKEN_PARSE_ERROR
  }

  private static class BuildOneTokenMethodResult {
    private final BuildOneTokenMethodResultType type;
    private final Object data;

    public BuildOneTokenMethodResult(BuildOneTokenMethodResultType type, Object data) {
      this.type = type;
      this.data = data;
    }

    public BuildOneTokenMethodResultType getType() {
      return type;
    }

    public static BuildOneTokenMethodResult generateTokenResult(Token data) {
      return new BuildOneTokenMethodResult(BuildOneTokenMethodResultType.TOKEN, data);
    }

    public Token getTokenData() {
      return (Token) data;
    }

    public static BuildOneTokenMethodResult generateAllTextHasBeenBuiltResult() {
      return new BuildOneTokenMethodResult(
          BuildOneTokenMethodResultType.ALL_TEXT_HAS_BEEN_BUILT, null);
    }

    public static BuildOneTokenMethodResult generateTokenParseErrorResult(
        BuildOneTokenMethodTokenGeneratorErrorData data) {
      return new BuildOneTokenMethodResult(BuildOneTokenMethodResultType.TOKEN_PARSE_ERROR, data);
    }

    public BuildOneTokenMethodTokenGeneratorErrorData
        getBuildOneTokenMethodTokenGeneratorErrorData() {
      return (BuildOneTokenMethodTokenGeneratorErrorData) data;
    }
  }

  public static class BuildOneTokenMethodTokenGeneratorErrorData {
    public final int start;
    public final int end;
    public final String errorText;

    public BuildOneTokenMethodTokenGeneratorErrorData(int start, int end, String errorText) {
      this.start = start;
      this.end = end;
      this.errorText = errorText;
    }
  }
}
